# Experiment

Experiments bao gồm các bài tập liên quan đến Part: D, F, G, H, I, J.

# 7. Part D — Experiment 1: Inspect the Sparse Representation
## 7.1. Dataset

In [1]:
import pandas as pd 

df = pd.read_json('/media/trong-thanh/Data/University/Fourth Year/NLP/experiments/ex1/data.json', lines = True)
df.head()

,text,timestamp,url
0,Beginners BBQ Class Taking Place in Missoula!\...,2019-04-25 12:57:54+00:00,https://klyq.com/beginners-bbq-class-taking-pl...
1,Discussion in 'Mac OS X Lion (10.7)' started b...,2019-04-21 10:07:13+00:00,https://forums.macrumors.com/threads/restore-f...
2,Foil plaid lycra and spandex shortall with met...,2019-04-25 10:40:23+00:00,https://awishcometrue.com/Catalogs/Clearance/T...
3,How many backlinks per day for new site?\nDisc...,2019-04-21 12:46:19+00:00,https://www.blackhatworld.com/seo/how-many-bac...
4,The Denver Board of Education opened the 2017-...,2019-04-20 14:33:21+00:00,http://bond.dpsk12.org/category/news/


## 7.2. Xây dựng pipeline

In [2]:
df['clean_text'] = (
    df['text']
    .fillna("")
    .str.lower()
    .str.replace(r'[^a-z0-9\s]', ' ', regex=True)
    .str.replace(r'\s+', ' ', regex=True)
    .str.strip()
)
df.head()

,text,timestamp,url,clean_text
0,Beginners BBQ Class Taking Place in Missoula!\...,2019-04-25 12:57:54+00:00,https://klyq.com/beginners-bbq-class-taking-pl...,beginners bbq class taking place in missoula d...
1,Discussion in 'Mac OS X Lion (10.7)' started b...,2019-04-21 10:07:13+00:00,https://forums.macrumors.com/threads/restore-f...,discussion in mac os x lion 10 7 started by ax...
2,Foil plaid lycra and spandex shortall with met...,2019-04-25 10:40:23+00:00,https://awishcometrue.com/Catalogs/Clearance/T...,foil plaid lycra and spandex shortall with met...
3,How many backlinks per day for new site?\nDisc...,2019-04-21 12:46:19+00:00,https://www.blackhatworld.com/seo/how-many-bac...,how many backlinks per day for new site discus...
4,The Denver Board of Education opened the 2017-...,2019-04-20 14:33:21+00:00,http://bond.dpsk12.org/category/news/,the denver board of education opened the 2017 ...


## 7.3. Kiểm tra kích thước

In [3]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
X_counter = vectorizer.fit_transform(df['clean_text'])
print(f'Shape: {X_counter.shape}')
print(X_counter)

Shape: (30000, 186582)
<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 4981696 stored elements and shape (30000, 186582)>
  Coords	Values
  (0, 25537)	1
  (0, 24928)	4
  (0, 39862)	4
  (0, 163491)	1
  (0, 130163)	1
  (0, 85196)	4
  (0, 110928)	1
  (0, 53714)	1
  (0, 185031)	5
  (0, 179407)	1
  (0, 168005)	5
  (0, 71973)	2
  (0, 26745)	2
  (0, 20945)	1
  (0, 104328)	1
  (0, 49782)	1
  (0, 181736)	5
  (0, 78127)	1
  (0, 166044)	4
  (0, 122420)	1
  (0, 135978)	1
  (0, 166665)	1
  (0, 121809)	1
  (0, 185083)	1
  (0, 33821)	1
  :	:
  (29999, 147475)	1
  (29999, 83782)	1
  (29999, 98770)	1
  (29999, 170702)	2
  (29999, 141337)	1
  (29999, 126750)	1
  (29999, 120446)	2
  (29999, 182339)	4
  (29999, 182341)	1
  (29999, 62303)	2
  (29999, 63822)	2
  (29999, 185062)	1
  (29999, 46009)	3
  (29999, 98342)	3
  (29999, 21344)	1
  (29999, 154670)	1
  (29999, 51076)	1
  (29999, 127744)	2
  (29999, 87041)	1
  (29999, 46591)	1
  (29999, 62295)	1
  (29999, 180317)	1
  (29999, 88579)	1
  (2999

Khi này:

N = 30000

V = 186582

## 7.4. Kiểm tra sparsity

In [4]:
N, V = X_counter.shape
print(f'N = {N}, V = {V}')

N = 30000, V = 186582


In [5]:
nnz = X_counter.nnz
total_elements = N * V
S = 1 - (nnz / total_elements)

print(f"Số lượng văn bản (N): {N:,}")
print(f"Kích thước từ điển (V): {V:,}")
print(f"Tổng số phần tử trong ma trận (N * V): {total_elements:,}")
print(f"Số phần tử khác 0 (nnz): {nnz:,}")
print(f"Tỷ lệ phần tử khác 0: {(nnz / total_elements) * 100:.4f}%")
print(f"Độ thưa Sparsity (S): {S:.6f} (tương đương {S * 100:.4f}%)")


Số lượng văn bản (N): 30,000
Kích thước từ điển (V): 186,582
Tổng số phần tử trong ma trận (N * V): 5,597,460,000
Số phần tử khác 0 (nnz): 4,981,696
Tỷ lệ phần tử khác 0: 0.0890%
Độ thưa Sparsity (S): 0.999110 (tương đương 99.9110%)


## 7.5. Inspect vocabulary

In [6]:
import numpy as np

feature_names = np.array(vectorizer.get_feature_names_out())
df_counts = np.array((X_counter > 0).sum(axis = 0)).flatten()

top_20 = np.argsort(df_counts)[::-1][:20]
print('Top 20 terns')
for rank, index in enumerate(top_20, 1):
    print(f'{rank:2d}. {feature_names[index]:<12}')

Top 20 terns
 1. the         
 2. and         
 3. to          
 4. of          
 5. in          
 6. for         
 7. is          
 8. with        
 9. on          
10. that        
11. this        
12. are         
13. it          
14. as          
15. at          
16. from        
17. be          
18. you         
19. by          
20. have        


In [7]:
from sklearn.feature_extraction.text import TfidfTransformer

tfidf_transformer = TfidfTransformer()
X_tfidf = tfidf_transformer.fit_transform(X_counter)
idfs = tfidf_transformer.idf_

In [8]:
top_20_idf = np.argsort(idfs)[::-1][:20]
print('Top 20 IDF')
for rank, index in enumerate(top_20_idf, 1):
    print(f'{rank:2d}. {feature_names[index]:<12}')

Top 20 IDF
 1. 00000       
 2. zzzzzzz333  
 3. zzzzzz      
 4. 00003       
 5. 000040      
 6. 00005       
 7. 0000856166  
 8. 0001042     
 9. 000116      
10. 00012       
11. 00015       
12. 00016       
13. 000165101   
14. 0002        
15. 00022       
16. 000226      
17. 000281      
18. zwyn        
19. zx636       
20. zx6rr       


In [9]:
tfidf_0 = X_tfidf.getrow(0).toarray().flatten()
top_20_tfidf_0 = np.argsort(tfidf_0)[::-1][:20]

for rank, index in enumerate(top_20_tfidf_0, 1):
    if tfidf_0[index] > 0:
        print(f'{rank:2d}. {feature_names[index]:<12}')

 1. bbq         
 2. class       
 3. meat        
 4. balay       
 5. kcbs        
 6. lonestar    
 7. will        
 8. missoula    
 9. apron       
10. smoker      
11. you         
12. timelines   
13. trimming    
14. spectators  
15. cost        
16. rangers     
17. 22nd        
18. beginner    
19. beginners   
20. culinary    


Trả lời:

- Từ xuất hienj nhiều trong toàn bộ corpus, khi đó DF = N, khi đó tỉ lệ N / DF = 1 => log (N / DF) = 0, khi đó xuất hiện nhiều lần trong văn bản, trọng số vẫn kéo gần về bằng 0.

- Một từ IDF cao => nó hiếm trong corpus. Đối với từ hiện đại trong văn bản, số lần xuất hiện bằng 0 => TF = 0 khi đó TF IDF = 0, từ đó IDF cao chỉ có giá trị cao trong chính tài liệu hiếm chứa nó.

# 9. Part F — Experiment 2: Preprocessing Ablation

In [10]:
import re
import numpy as np
import scipy.sparse as sp
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.metrics.pairwise import cosine_similarity
from tokenizers import Tokenizer

In [11]:
import pandas as pd 
import time

PATH = '/media/trong-thanh/Data/University/Fourth Year/NLP/experiments/ex1/data.json'
df = pd.read_json(PATH, lines = True)
texts = df['text'].fillna('').tolist()

In [12]:
for index, i in enumerate(texts[:10]):
    print(f'Text {index}')
    print(i)

Text 0
Beginners BBQ Class Taking Place in Missoula!
Do you want to get better at making delicious BBQ? You will have the opportunity, put this on your calendar now. Thursday, September 22nd join World Class BBQ Champion, Tony Balay from Lonestar Smoke Rangers. He will be teaching a beginner level class for everyone who wants to get better with their culinary skills.
He will teach you everything you need to know to compete in a KCBS BBQ competition, including techniques, recipes, timelines, meat selection and trimming, plus smoker and fire information.
The cost to be in the class is $35 per person, and for spectators it is free. Included in the cost will be either a t-shirt or apron and you will be tasting samples of each meat that is prepared.
Text 1
Discussion in 'Mac OS X Lion (10.7)' started by axboi87, Jan 20, 2012.
I've got a 500gb internal drive and a 240gb SSD.
When trying to restore using disk utility i'm given the error "Not enough space on disk ____ to restore"
But I shouldn

## 9.1. Pipeline A — Minimal

In [13]:
vector_1 = CountVectorizer(lowercase=True, token_pattern=r'(?u)\b\w+\b')
X_1 = vector_1.fit_transform(texts)

vocab_1 = len(vector_1.vocabulary_)
avg_tok_a = float(X_1.sum(axis=1).mean())
sparsity_a = 1.0 - (X_1.nnz / (X_1.shape[0] * X_1.shape[1]))


In [14]:
print(f'Vocabulary size: {vocab_1}')
print(f'Average token/docs: {avg_tok_a}')
print(f'Matrix shape: {X_1.shape}')
print(f'Matrix sparsity (S): {sparsity_a:.6f} ({sparsity_a * 100:.4f}%)')

Vocabulary size: 193837
Average token/docs: 369.6963666666667
Matrix shape: (30000, 193837)
Matrix sparsity (S): 0.999122 (99.9122%)


Pipeline A khá đơn giản Raw -> Lowercase -> Tokenization

Số từ vựng rất lớn so với 30000 văn bản, khoảng ~6 từ vựng / văn bản tức mỗi văn bản đóng góp trung bình 6 từ vựng.

Mức độ thưa rất lớn, spa 99.9122% tức tỉ lệ % các ô có giá trị bằng 0 trong các văn bản rất lớn. Tỷ lệ ô khác 0 = 100% − Sparsity = 100% − 99.9122% = 0.0878% nên mỗi văn bản trung bình chỉ chứa khoảng 170 từ vựng riêng biệt (không tính trùng lặp) trên tổng số 193.837 từ trong toàn bộ từ điển.

## 9.2. Pipeline B — Normalized

In [15]:
import string 

regex = re.compile(f'[{re.escape(string.punctuation)}]')

def normalize(text):
    return regex.sub(' ', text.lower())

cleaned_texts_2 = [normalize(t) for t in texts]    

In [16]:
vector_2 = CountVectorizer(
    stop_words='english',         
    token_pattern=r'(?u)\b\w+\b'  
)
X_2 = vector_2.fit_transform(cleaned_texts_2)

In [17]:
vocab_b = len(vector_2.vocabulary_)
avg_tok_b = float(X_2.sum(axis=1).mean())
sparsity_b = 1.0 - (X_2.nnz / (X_2.shape[0] * X_2.shape[1]))

In [18]:
print(f'Vocabulary size: {vocab_b}')
print(f'Average token/docs: {avg_tok_b}')
print(f'Matrix shape: {X_2.shape}')
print(f'Matrix sparsity (S): {sparsity_b:.6f} ({sparsity_b * 100:.4f}%)')

Vocabulary size: 192764
Average token/docs: 197.4658
Matrix shape: (30000, 192764)
Matrix sparsity (S): 0.999351 (99.9351%)


Pipeline B thêm 2 bước so với A chuẩn hóa dấu câu và loại bỏ stop words.

Số từ vựng giảm từ 193.837 xuống 192.764 do loại bỏ stop words và làm sạch các từ dính dấu câu.

Số lượng token trung bình mỗi văn bản giảm mạnh từ ~370 xuống ~197 token (giảm gần 50%), cho thấy stop words chiếm phần lớn số lượng từ trong văn bản.

Độ thưa tăng lên 99.9351%. Tỷ lệ ô khác 0 còn: 100% − 99.9351% = 0.0649%, tức mỗi văn bản giờ chỉ còn trung bình khoảng 125 từ vựng riêng biệt (giảm từ 170 ở Pipeline A do các từ phổ biến đã bị lọc bỏ).


## 9.3. Pipeline C — Extended

In [19]:
tok = Tokenizer.from_pretrained('bert-base-uncased')
encs = tok.encode_batch(texts)
vocab_3 = tok.get_vocab_size()

lengths_c = [len(e.ids) for e in encs]
avg_tok_c = float(np.mean(lengths_c))

/home/trong-thanh/miniconda3/envs/ml/lib/python3.14/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [20]:
indptr = [0]
indices = []
data = []
for e in encs:
    u_ids, counts = np.unique(e.ids, return_counts=True)
    indices.extend(u_ids)
    data.extend(counts)
    indptr.append(len(indices))


In [21]:
X_3 = sp.csr_matrix((data, indices, indptr), shape=(len(texts), vocab_3), dtype=np.int32)
sparsity_c = 1.0 - (X_3.nnz / (X_3.shape[0] * X_3.shape[1]))


In [22]:
print(f'Vocabulary size: {vocab_3}')
print(f'Average token/docs: {avg_tok_c}')
print(f'Matrix shape: {X_3.shape}')
print(f'Matrix sparsity (S): {sparsity_c:.6f} ({sparsity_c * 100:.4f}%)')


Vocabulary size: 30522
Average token/docs: 467.0992
Matrix shape: (30000, 30522)
Matrix sparsity (S): 0.993621 (99.3621%)


Pipeline C sử dụng subword tokenization (WordPiece) với từ điển cố định của mô hình BERT.

Từ điển cố định nhỏ hơn rất nhiều so với A và B, chỉ có 30.522 từ vựng. 

Số token trung bình mỗi văn bản tăng lên ~467 token vì các từ dài hoặc từ hiếm bị bẻ nhỏ thành nhiều subword.

Độ thưa giảm xuống còn 99.3621% (ít thưa nhất trong cả 3 pipeline). Tỷ lệ ô khác 0 là 0.6379%, cao gấp gần 10 lần Pipeline B vì từ điển nhỏ và mỗi văn bản chứa nhiều subword lặp lại.


## 9.4. So sánh

In [23]:
queries = [
    "medical image classification",
    "transformer language model",
    "deep learning healthcare",
    "natural language processing",
    "myocardial infarction cardiology",
    "covid-19 pneumonia coronavirus"
]

In [24]:
tokens_a = [w.lower() for q in queries for w in re.findall(r'\b\w+\b', q)]
oov_a = sum(1 for w in tokens_a if w not in vector_1.vocabulary_) / len(tokens_a) * 100

stop_b = vector_2.get_stop_words()
tokens_b = [w for w in tokens_a if w not in stop_b]
oov_b = sum(1 for w in tokens_b if w not in vector_2.vocabulary_) / len(tokens_b) * 100

unk_id = tok.token_to_id("[UNK]")
encs_q = tok.encode_batch(queries)
total_sub = sum(len(e.ids) for e in encs_q)
oov_c = sum(e.ids.count(unk_id) for e in encs_q) / total_sub * 100


In [25]:
tfidf_1 = TfidfTransformer().fit_transform(X_1)
tfidf_2 = TfidfTransformer().fit_transform(X_2)
tfidf_3 = TfidfTransformer().fit_transform(X_3)
def get_top5_sim(queries, vec, tfidf_mat, is_subword=False):
    res = []
    for q in queries:
        if not is_subword:
            q_v = TfidfTransformer().fit(tfidf_mat).transform(vec.transform([q]))
        else:
            e = tok.encode(q)
            u, cnt = np.unique(e.ids, return_counts=True)
            csr = sp.csr_matrix((cnt, ([0]*len(u), u)), shape=(1, vocab_3), dtype=np.float32)
            q_v = TfidfTransformer().fit(tfidf_mat).transform(csr)
        s = cosine_similarity(q_v, tfidf_mat).flatten()
        res.append(np.mean(np.sort(s)[::-1][:5]))
    return float(np.mean(res))


In [26]:
sim_a = get_top5_sim(queries, vector_1, tfidf_1)
sim_b = get_top5_sim(queries, vector_2, tfidf_2)
sim_c = get_top5_sim(queries, tok, tfidf_3, is_subword=True)
df_compare = pd.DataFrame({
    "Metric": [
        "Vocabulary size",
        "Average tokens/document",
        "Matrix sparsity",
        "OOV rate",
        "Search performance (Top 5 sim)"
    ],
    "Pipeline A": [f"{vocab_1:,}", f"{avg_tok_a:.2f}", f"{sparsity_a:.6f}", f"{oov_a:.2f}%", f"{sim_a:.4f}"],
    "Pipeline B": [f"{vocab_b:,}", f"{avg_tok_b:.2f}", f"{sparsity_b:.6f}", f"{oov_b:.2f}%", f"{sim_b:.4f}"],
    "Pipeline C": [f"{vocab_3:,}", f"{avg_tok_c:.2f}", f"{sparsity_c:.6f}", f"{oov_c:.2f}%", f"{sim_c:.4f}"]
})
display(df_compare)


,Metric,Pipeline A,Pipeline B,Pipeline C
0,Vocabulary size,"193,837","192,764","30,522"
1,Average tokens/document,369.70,197.47,467.10
2,Matrix sparsity,0.999122,0.999351,0.993621
3,OOV rate,5.26%,5.26%,0.00%
4,Search performance (Top 5 sim),0.2499,0.2651,0.2704


## 9.5. Câu hỏi phân tích

### 1. Lowercasing làm thay đổi vocabulary như thế nào?

- Thực nghiệm cho thấy khi chưa lowercasing, từ điển có 240.792 từ vựng. Sau khi lowercasing, từ điển giảm xuống còn 193.837 từ (giảm 46.955 từ, tương đương 19.50%).
- Việc giảm này giúp gom các biến thể, nhưng làm mất phân biệt thực thể: trong dữ liệu gốc tồn tại đồng thời cả hai từ mang nghĩa khác nhau như `US` (nước Mỹ) và `us` (chúng tôi), `Apple` (công ty) và `apple` (quả táo), sau khi lowercasing chúng bị gộp làm một.


In [27]:
text_sample = texts[9]
upper_tokens = [w for w in re.findall(r'\b[A-Z]{2,}\b', text_sample)[:3]]
print("Từ viết hoa gốc trong văn bản:", upper_tokens)
print("Từ sau khi qua vector_1:", [w.lower() for w in upper_tokens])


Từ viết hoa gốc trong văn bản: ['US', 'US', 'US']
Từ sau khi qua vector_1: ['us', 'us', 'us']


### 2. Stopword removal có luôn cải thiện representation không?

Việc bỏ không phải lúc nào cũng tốt, bỏ stopword giúp văn bản ngắn gọn, tập trung vào từ khóa chính, nhưng sẽ làm mất nghĩa của các câu ngắn, câu thành ngữ (như "to be or not to be") hoặc các cụm từ quan trọng.


In [28]:
q_stop = "to be or not to be"
print(f"Số từ tìm thấy ở Pipeline A: {vector_1.transform([q_stop]).nnz} từ")
print(f"Số từ tìm thấy ở Pipeline B: {vector_2.transform([q_stop]).nnz} từ")


Số từ tìm thấy ở Pipeline A: 4 từ
Số từ tìm thấy ở Pipeline B: 0 từ


### 3. Việc loại punctuation có thể làm mất thông tin gì?

Mất số thập phân (3.14 thành 3 và 14), tên viết tắt, tên công nghệ (C++, C# thành C), email, đường link và sắc thái cảm xúc (!, ?).


Câu gốc: Mac OS X Lion (10.7) and covid-19

Tokens thu được: ['mac', 'os', 'x', 'lion', '10', '7', 'and',  'covid', '19']

10.7 bị tách thành: ['10', '7']

covid-19 bị tách thành: ['covid', '19']

### 4. Pipeline nào tạo ra sparse matrix nhất?

Pipeline B thưa nhất (99.9351%) vì vừa có kích thước từ điển lớn vừa bỏ đi các từ xuất hiện nhiều là stop words. Pipeline C ít thưa nhất (99.3621%) do từ điển cố định nhỏ.

### 5. Pipeline nào cho search tốt nhất?

Pipeline B cho kết quả điểm tương đồng cao nhất với các query từ khóa chuẩn vì không bị nhiễu bởi stopword. Pipeline C tốt nhất khi gặp từ mới, từ hiếm hoặc từ sai chính tả vì không bị OOV.

In [29]:
tu = "myocardial"
print("Subword:", tok.encode(tu).tokens)


Subword: ['[CLS]', 'my', '##oca', '##rdial', '[SEP]']


### 6. Search tốt hơn có đồng nghĩa với vocabulary nhỏ hơn không?
Không đồng nghĩa. Kích thước từ điển không quyết định chất lượng tìm kiếm:
- Pipeline A và B, từ điển của B chỉ giảm ít (~0.5%) nhưng search tốt hơn rõ rệt nhờ loại bỏ stopword gây nhiễu.
- Nếu dùng từ điển nhỏ, các từ khóa chuyên ngành hiếm (như myocardial, infarction) sẽ bị OOV hoàn toàn, khiến hệ thống không tìm được.

- Pipeline C từ điển nhỏ nhất (30.522 từ) nhưng bẻ từ thành subword, dễ gây trùng khớp giả (ví dụ myocardial bị bẻ có chứa mảnh "my", vô tình khớp với câu chứa đại từ "my").

In [ ]:
print("Index của 'myocardial' trong Pipeline B:", vector_2.vocabulary_.get("myocardial"))
print("Subwords 'myocardial':", tok.encode("myocardial").tokens[1:-1])
print("Subwords 'this is my cat':", tok.encode("this is my cat").tokens[1:-1])

Index của 'myocardial' trong Pipeline B: 115869
Subwords 'myocardial': ['my', '##oca', '##rdial']
Subwords 'this is my cat': ['this', 'is', 'my', 'cat']


# 10. Part G — Application: Build a Document Search Engine
## 10.1. Bài toán & 10.2. Pipeline

In [31]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.metrics.pairwise import cosine_similarity

df = pd.read_json('/media/trong-thanh/Data/University/Fourth Year/NLP/experiments/ex1/data.json', lines=True)
texts = df['text'].fillna('').tolist()
print(f'So luong documents: {len(texts):,}')


So luong documents: 30,000


In [32]:
vectorizer = CountVectorizer(lowercase=True, stop_words='english', token_pattern=r'(?u)\b\w+\b')
X = vectorizer.fit_transform(texts)

tfidf_trans = TfidfTransformer()
X_tfidf = tfidf_trans.fit_transform(X)
print(f'X_tfidf shape: {X_tfidf.shape}')


X_tfidf shape: (30000, 193521)


In [33]:
def search(query, k=5):
    q_vec = tfidf_trans.transform(vectorizer.transform([query]))
    sims = cosine_similarity(q_vec, X_tfidf).flatten()
    top_idx = np.argsort(sims)[::-1][:k]
    
    rows = []
    for rank, idx in enumerate(top_idx, 1):
        rows.append({
            'Rank': rank,
            'Doc ID': idx,
            'Similarity': round(float(sims[idx]), 4),
            'Preview': texts[idx].replace('\n', ' ')[:100] + '...'
        })
    return pd.DataFrame(rows)


## 10.3. Query examples & 10.4. Kết quả cần hiển thị

In [34]:
queries = [
    "medical image classification",
    "transformer language model",
    "deep learning healthcare",
    "natural language processing"
]

for q in queries:
    print(f"Query: {q}")
    print(search(q).to_string(index=False))
    print()


Query: medical image classification
 Rank  Doc ID  Similarity                                                                                                 Preview
    1   18971      0.4237 The new RTS Environmental Classification system (RTS GLT) is designed for parties who are commission...
    2    8527      0.3537 History of maize classification. How races used in classification. Geographical distribution. Existi...
    3   19908      0.2609 Download League Of Legends Wallpapers in high-quality for your desktop and smart-phone in wide-scree...
    4   12658      0.2490 This guidance is for pharmacists who handle, use and sell/supply medical devices. What to do with me...
    5   15682      0.2486 - Group Image: Provided functionality of group's image, user can now view and change the group's ima...

Query: transformer language model
 Rank  Doc ID  Similarity                                                                                                 Preview
    1   27936      

### Nhận xét
- Tốc độ tìm kiếm nhanh nhờ nhân vô hướng ma trận thưa.
- Các tài liệu đứng đầu đều chứa đúng các từ khóa truy vấn.
- Hạn chế: Không bắt được từ đồng nghĩa nếu văn bản không chứa đúng từ vựng của query.

# 11. Part H — Evaluation
## 11.1 - 11.4. Quantitative Evaluation (Precision@5, Recall@5, MRR)

In [35]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.metrics.pairwise import cosine_similarity

df = pd.read_json('/media/trong-thanh/Data/University/Fourth Year/NLP/experiments/ex1/data.json', lines=True)
texts = df['text'].fillna('').tolist()

vectorizer = CountVectorizer(lowercase=True, stop_words='english', token_pattern=r'(?u)\b\w+\b')
X = vectorizer.fit_transform(texts)
tfidf_trans = TfidfTransformer()
X_tfidf = tfidf_trans.fit_transform(X)

In [36]:
eval_set = {
    "medical image classification": [18971, 12658],
    "deep learning healthcare": [6123, 8270],
    "natural language processing": [5912, 23722],
    "transformer language model": [18339],
    "heart attack treatment": [799, 1422]
}

def evaluate(queries_dict, k=5):
    rows = []
    for q, rel in queries_dict.items():
        q_vec = tfidf_trans.transform(vectorizer.transform([q]))
        sims = cosine_similarity(q_vec, X_tfidf).flatten()
        ret = [int(x) for x in np.argsort(sims)[::-1][:k]]
        
        hits = [d for d in ret if d in rel]
        p_k = len(hits) / k
        r_k = len(hits) / len(rel)
        
        rr = 0.0
        for rank, d in enumerate(ret, 1):
            if d in rel:
                rr = 1.0 / rank
                break
                
        rows.append({
            'query': q,
            'p_at_5': round(p_k, 4),
            'r_at_5': round(r_k, 4),
            'mrr': round(rr, 4),
            'retrieved': ret,
            'relevant': rel
        })
    return pd.DataFrame(rows)

results = evaluate(eval_set)
display(results)
print(f"Mean P@5: {results['p_at_5'].mean():.4f}")
print(f"Mean R@5: {results['r_at_5'].mean():.4f}")
print(f"MRR: {results['mrr'].mean():.4f}")


,query,p_at_5,r_at_5,mrr,retrieved,relevant
0,medical image classification,0.4,1.0,1.0,"[18971, 8527, 19908, 12658, 15682]","[18971, 12658]"
1,deep learning healthcare,0.2,0.5,0.5,"[7564, 6123, 11979, 11119, 9252]","[6123, 8270]"
2,natural language processing,0.0,0.0,0.0,"[25428, 8705, 24482, 4075, 701]","[5912, 23722]"
3,transformer language model,0.0,0.0,0.0,"[27936, 25428, 24482, 4075, 701]",[18339]
4,heart attack treatment,0.4,1.0,1.0,"[799, 1422, 11245, 22876, 16626]","[799, 1422]"


Mean P@5: 0.2000
Mean R@5: 0.5000
MRR: 0.5000


In [37]:
results.to_csv('results.csv', index=False)

# 12. Part I — Error Analysis

Medical image classification và heart attack treatment đạt kết quả cao nhất với MRR và R@5 đều bằng 1.0. Các tài liệu đứng đầu như số 18971 và 799 đều chứa đúng các từ khóa quan trọng mang giá trị IDF lớn như classification hay cụm từ heart attack. 

Trùng khớp từ vựng trực tiếp giúp hệ thống xếp đúng toàn bộ tài liệu y khoa cần tìm vào nhóm dẫn đầu, dù vẫn bị lọt vào vài văn bản nhiễu do trùng lặp đơn lẻ từng từ.


In [38]:
good_queries = ["medical image classification", "heart attack treatment"]
for q in good_queries:
    q_vec = tfidf_trans.transform(vectorizer.transform([q]))
    sims = cosine_similarity(q_vec, X_tfidf).flatten()
    top_idx = int(np.argsort(sims)[::-1][0])
    print(f"Query '{q}' -> Top 1 Doc {top_idx} | Similarity {sims[top_idx]:.4f}")
    print(f"From {texts[top_idx].replace(chr(10), ' ')[:100]}...\n")


Query 'medical image classification' -> Top 1 Doc 18971 | Similarity 0.4237
From The new RTS Environmental Classification system (RTS GLT) is designed for parties who are commission...

Query 'heart attack treatment' -> Top 1 Doc 799 | Similarity 0.3962
From Can You Have A Heart Attack With A Pacemaker? A pacemaker is not a cure. It will not prevent or stop...



Ngược lại, hai truy vấn natural language processing và deep learning healthcare thất bại do hạn chế của mô hình túi từ. 

Khi tìm kiếm hệ thống bị điểm 0 hoàn toàn vì bẻ gãy cụm từ thành các từ rời rạc, dẫn đến việc xếp đầu các bài viết về đổi ngôn ngữ app Facebook và an toàn chế biến thực phẩm. 

Với câu truy vấn về học sâu y tế, một bảng chia thì động từ lặp đi lặp lại chữ learning đã leo lên vị trí số một do tần suất TF tăng tuyến tính không bão hòa, đẩy lùi các bài báo khoa học thực sự.


In [39]:
bad_queries = ["natural language processing", "deep learning healthcare"]
for q in bad_queries:
    q_vec = tfidf_trans.transform(vectorizer.transform([q]))
    sims = cosine_similarity(q_vec, X_tfidf).flatten()
    top_idx = int(np.argsort(sims)[::-1][0])
    print(f"Query '{q}' -> Top 1 Doc {top_idx} | Similarity {sims[top_idx]:.4f}")
    print(f"From {texts[top_idx].replace(chr(10), ' ')[:100]}...\n")


Query 'natural language processing' -> Top 1 Doc 25428 | Similarity 0.3893
From Note: If you're on an iPhone, you cannot change the language of Facebook through the mobile app. Ins...



Query 'deep learning healthcare' -> Top 1 Doc 7564 | Similarity 0.3094
From would I have learned, learnt? would you have learned, learnt? would he have learned, learnt? would w...



### Failure case quan trọng nhất

Trường hợp thất bại là truy vấn heart attack treatment đối chiếu với bài viết số 8343 về liệu pháp nhồi máu cơ tim myocardial infarction therapy. Hai khái niệm hoàn toàn đồng nghĩa trong y khoa, điểm tương đồng cosine vẫn bằng đúng 0.0000 vì chúng không chung từ vựng nào. 

 TF-IDF khiến hệ thống thất bại trước hiện tượng từ đồng nghĩa và từ đa nghĩa. Điều này đòi hỏi chúng ta phải chuyển dịch sang các không gian biểu diễn ngữ nghĩa dày đặc như Word2Vec hay BERT để lấy được tương đồng ngữ nghĩa thực sự.

In [41]:
doc_fail = next(i for i, t in enumerate(texts) if 'myocardial infarction' in t.lower() and 'therapy' in t.lower() and 'heart' not in t.lower())

q_fail = 'heart attack treatment'
q_fail_vec = tfidf_trans.transform(vectorizer.transform([q_fail]))
sim_fail = cosine_similarity(q_fail_vec, X_tfidf[doc_fail:doc_fail+1])[0, 0]

print(f"Doc ID {doc_fail}")
print(f"From {texts[doc_fail][:100]}...")
print(f"Similarity with '{q_fail}' la {sim_fail:.4f}")


Doc ID 8343
From Background: Statin therapy is associated with improved survival in patients at high risk for cardiov...
Similarity with 'heart attack treatment' la 0.0000


# 13. Part J — From Failure to the Next NLP Representation

Thực nghiệm cho thấy hạn chế lớn nhất của mô hình TF-IDF là phụ thuộc hoàn toàn vào việc trùng khớp từ vựng (lexical matching). Khi hai văn bản diễn đạt cùng một ý nhưng dùng các từ đồng nghĩa khác nhau, hệ thống sẽ trả về độ tương đồng bằng 0.

Để giải quyết vấn đề này, ta cần dựa trên giả thuyết phân phối, những từ xuất hiện trong các ngữ cảnh tương tự nhau thì thường có ý nghĩa tương tự nhau.

Giả thuyết này mở ra hướng tiếp cận mới: thay vì biểu diễn mỗi từ bằng một chiều rời rạc trong không gian thưa, ta ánh xạ chúng thành các vector dày đặc (dense embeddings) trong không gian ít chiều hơn. Khi đó, các khái niệm liên quan như heart attack và myocardial infarction sẽ nằm gần nhau về mặt hình học.

Lộ trình phát triển tự nhiên của biểu diễn ngữ nghĩa sẽ đi qua các giai đoạn:
TF-IDF -> Distributional representation -> Word Embedding (Word2Vec, FastText) -> Contextual Embedding (BERT, RoBERTa) -> Transformer.

# 15. Learning Check

**Question 1:** Tại sao TF-IDF tạo ra sparse representation?

Trả lời:

Vector biểu diễn của mỗi văn bản có số chiều bằng bộ kích thước của tập từ điển, tuy nhiên một văn bản thông thường chỉ sử dụng một lượng từ rất nhỏ, điều này dẫn đến hơn 99.93% các vị trí tọa độ của vector đều mang giá trị 0, chỉ có một phần các phần tử rất nhỏ mang giá trị khác 0 tạo nên ma trận thưa.

**Question 2:** Tại sao một term xuất hiện trong hầu hết documents có IDF thấp?

Trả lời:

Theo công thức IDF(t) = log(N / df(t)), khi từ điển xuất hiện trong hầu hết văn bản thì df(t) xấp xỉ bằng N => tỷ số N / df(t) ~1, mà log(1) = 0. Về ý nghĩa, những từ xuất hiện ở mọi nơi như the, is, a,... không có giá trị phân biệt nội dung giữa văn bản với văn bản, nên trọng số của chúng bị kéo về gần giá trị bằng 0.

**Question 3:** Tại sao một term có IDF cao chưa chắc có TF-IDF cao trong một document?

Trả lời:

Công thức tính TF-IDF = TF * IDF, một từ có idf cao => là từ hiếm gặp trong bộ dữ liệu. Tuy nhiên nếu xét văn bản đang xét hoàn toàn không chứa từ này thì tf = 0 khéo theo tfidf = 0.

Một từ có tfidf cao khi vừa là từ hiếm, trong toàn bộ tập dữ liệu, vừa xuất hiện nhiều lần trong chính văn bản đó.

**Question 4:** Tại sao cosine similarity phù hợp với document vectors?

Trả lời:

Cosine Similarity đo góc lệch giữa hướng 2 vector thay vì đo khoảng cách chiều dài. Nhờ việc chia cho độ dài vector, phép đo làm mất đi độ chênh lệch về độ dài văn bản => dễ dàng để có thể so sánh và đánh gía.

**Question 5:** Tại sao preprocessing có thể thay đổi search result?

Trả lời:

Tiền xử lý tác động vào tập từ vựng, số lượng từ, tần xuất tf và trọng số idf:

- Với chuyển chữ thường, giúp gộp các từ viết hoa thành viết thường tránh làm phân mảnh từ điển.

- Loại bỏ từ  dừng, bỏ các từ xuất hiện quá nhiều, giúp giảm nhiễu và tập trung vào các từ nội dung chính.

- Loại bỏ dấu câu có thể làm mất đi một số nội dung quan trọng nên việc này cần làm rất cẩn thận.

- Tách từ theo từ đơn hay tách theo mảnh từ quyết định việc hệ thống có bị gặp từ lạ, phân mảnh hay khớp giả hay không.

**Question 6:** Một failure case của TF-IDF search mà em quan sát được là gì?

Trả lời: 

Hiện tượng không trùng khớp mặt chữ giữa các từ đồng nghĩa ví dụ:
- Câu truy vấn 'heart attack treatment'
- Văn bản cần tìm 'myocardial infarction therapy'

Hai câu này cùng 1 nghĩa y khoa nhưng vì không có chung chữ cái nào nên độ tương đồng bằng 0 => bỏ lỡ tài liệu.

**Question 7:** Failure case đó gợi ý nhu cầu về representation nào tiếp theo?

Trả lời:

Cho thấy cần phải chuyển từ việc khớp chữ sang biểu diễn ngữ cảnh dày đặc (Dense Semantic Representation) dựa trên giả thuyết phân phối các từ hay xuất hiện trong cùng ngữ xảnh sẽ  hoặc có ý nghĩa gần tương tự nhau được ánh xạ thành các vector số thực có số chiều nhỏ hơn và nằm gần nhau trong không gian.

Hướng phát triển:

tf-idf -> Word Embedding -> Context Embedding.